### PR TITLE
bpo-32554: Deprecate hashing arbitrary types in random.seed()

### PR DESCRIPTION
**Pull request by rhettinger**
https://bugs.python.org/issue32554

### CODE REVIEW

File: Doc/library/random.rst [hunk 1]
```diff
@@ -86,6 +86,11 @@ Bookkeeping functions
    .. versionchanged:: 3.2
       Moved to the version 2 scheme which uses all of the bits in a string seed.
 
+   .. deprecated:: 3.9
+      In the future, the *seed* must be one of the following types:
+      :class:`NoneType`, :class:`int`, :class:`float`, :class:`str`,
```
Comment (serhiy-storchaka) [inline Doc/library/random.rst]: There is no an entry for `NoneType`, so the link will not work.

File: Lib/random.py [hunk 2]
```diff
@@ -143,12 +146,19 @@ def seed(self, a=None, version=2):
...
+        elif version == 2 and isinstance(a, (str, bytes, bytearray)):
             if isinstance(a, str):
                 a = a.encode()
             a += _sha512(a).digest()
             a = int.from_bytes(a, 'big')
 
+        elif not isinstance(a, (type(None), int, float, str, bytes, bytearray)):
+            _warn('Seeding based on hashing is deprecated\n'
+                  'since Python 3.9 and will be removed in a subsequent '
+                  'version. The only \nsupported seed types are: None, '
```
Comment (serhiy-storchaka) [inline Lib/random.py]: I would break a string literal and a line after `\n` for readability

File: Doc/library/random.rst (hunk 1 above)
Comment (serhiy-storchaka) [reply Doc/library/random.rst]: Is `float` because of `seed(time.time())`?

**Review by serhiy-storchaka — COMMENTED** [review]
I am just interesting, why implement a warning in `random.Random.seed()` instead of `_random.Random.seed()`?

File: Doc/whatsnew/3.9.rst [hunk 3]
```diff
@@ -169,6 +169,12 @@ Deprecated
   of Python. For the majority of use cases users can leverage the Abstract Syntax
   Tree (AST) generation and compilation stage, using the :mod:`ast` module.
 
+* The :mod:`random` module currently accepts any hashable type as a
+  possible seed value.  Unfortunately, some of those types are not
+  guaranteed to have a deterministic hash value.  After Python 3.9,
+  the module will restrict its seeds to *None*, :class:`int`,
+  :class:`float`, :class:`str`, :class:`bytes`, and :class:`bytearray'.
```
Comment (serhiy-storchaka) [inline Doc/whatsnew/3.9.rst]:
```suggestion
  :class:`float`, :class:`str`, :class:`bytes`, and :class:`bytearray`.
```

File: Doc/library/random.rst (hunk 1 above)
Comment (rhettinger) [reply Doc/library/random.rst]: Yes.

**Comment by serhiy-storchaka** [comment]
It is unrelated. `hashlib` does not allow you to make a hash from arbitrary object.

**Pull request merged**